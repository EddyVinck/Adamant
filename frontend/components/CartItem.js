import React from "react";
import formatMoney from "../lib/formatMoney";
import styled from "styled-components";
import PropTypes from "prop-types";
import RemoveFromCart from "./RemoveFromCart";

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;

  h3,
  p {
    margin: 0;
  }
`;

const ImagePlaceHolder = styled.div`
  width: 100px;
  margin-right: 10px;

  img {
    max-width: 100%;
  }
`;

export default function CartItem({ cartItem }) {
  // First check if that item exists (could have been deleted)
  if (!cartItem.item)
    return (
      <CartItemStyles>
        <ImagePlaceHolder />
        <p>This item has been removed.</p>
        <RemoveFromCart id={cartItem.id} />
      </CartItemStyles>
    );
  const { id, title, description, image, price } = cartItem.item;
  const { quantity } = cartItem;
  return (
    <CartItemStyles>
      <ImagePlaceHolder>
        <img src={image} alt={title} />
      </ImagePlaceHolder>
      <div className="cart-item-details">
        <h3>{title}</h3>
        <p>
          {formatMoney(quantity * price)}
          {" - "}
          <em>
            {quantity} &times; {formatMoney(price)} each
          </em>
        </p>
      </div>
      <RemoveFromCart id={cartItem.id} />
    </CartItemStyles>
  );
}

CartItem.propTypes = {
  cartItem: PropTypes.object.isRequired
};
