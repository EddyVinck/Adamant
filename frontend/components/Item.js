import React, { Component } from "react";
import Link from "next/link";
import { PropTypes } from "prop-types";
import Title from "./styles/Title";
import ItemStyles from "./styles/ItemStyles";
import formatMoney from "../lib/formatMoney";
import DeleteItem from "./DeleteItem";
import AddToCart from "./AddToCart";

export default class Item extends Component {
  static propTypes = {
    item: PropTypes.object.isRequired
  };

  render() {
    const { item } = this.props;
    const descriptionLength = 60;
    let description =
      item.description.length <= descriptionLength
        ? item.description
        : item.description.substring(0, descriptionLength) + "...";
    return (
      <ItemStyles>
        <Link
          href={{
            pathname: "/item",
            query: { id: item.id }
          }}
        >
          <a>
            {item.image && (
              <div className="img-wrapper">
                <img src={item.image} alt={item.title} />
              </div>
            )}
          </a>
        </Link>
        <div className="item__details">
          <Link
            href={{
              pathname: "/item",
              query: { id: item.id }
            }}
          >
            <a>
              <Title>{item.title}</Title>
              <p>{description}</p>
              <span className="price">{formatMoney(item.price)}</span>
            </a>
          </Link>
        </div>
        <div className="buttonList">
          <AddToCart id={item.id} />
          <Link href={{ pathname: "update", query: { id: item.id } }}>
            <a>Edit</a>
          </Link>
          <DeleteItem id={item.id}>Delete</DeleteItem>
        </div>
      </ItemStyles>
    );
  }
}
