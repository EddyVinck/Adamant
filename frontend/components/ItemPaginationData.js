import React, { Component } from "react";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import Error from "./ErrorMessage";
import { perPage as itemsPerPage } from "../config";
import PropTypes from "prop-types";

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

// This is a render prop provider component
class PaginationData extends Component {
  static propTypes = {
    loading: PropTypes.element,
    pathname: PropTypes.string,
    page: PropTypes.number.isRequired
  };
  static defaultProps = {
    loading: null,
    pathname: "items"
  };
  render() {
    return (
      <Query query={PAGINATION_QUERY}>
        {({ error, loading, data }) => {
          const loadingElement = this.props.loading || <p>Loading...</p>;
          if (error) return <Error error={error} />;
          if (loading) return loadingElement;
          const { count: itemCount } = data.itemsConnection.aggregate;
          const { page: currentPage, pathname } = this.props;
          const pagesCount = Math.ceil(itemCount / itemsPerPage);
          const prevPage = {
            pathname,
            query: { page: currentPage - 1 }
          };
          const nextPage = {
            pathname,
            query: { page: currentPage + 1 }
          };
          const prevDisabled = currentPage <= 1;
          const nextDisabled = currentPage >= pagesCount;
          return (
            <>
              {this.props.children({
                loading,
                currentPage,
                itemCount,
                pagesCount,
                prevPage,
                nextPage,
                prevDisabled,
                nextDisabled
              })}
            </>
          );
        }}
      </Query>
    );
  }
}

export default PaginationData;
export { PAGINATION_QUERY };
