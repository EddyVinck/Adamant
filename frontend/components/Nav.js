import Link from "next/link";
import { Mutation } from "react-apollo";
import { TOGGLE_CART_MUTATION } from "./Cart";
import NavStyles from "./styles/NavStyles";
import User from "./User";
import SignoutButton from "./Signout";
import CartCount from "./CartCount";
import styled from "styled-components";
import Search from "./Search";

const BottomBar = styled.div`
  display: flex;
  background: black;
  @media (min-width: ${props => props.theme.breakpoints.large}) {
    justify-content: flex-end;
  }
`;
const TopBar = styled(BottomBar)`
  margin-bottom: 1rem;
  @media (min-width: ${props => props.theme.breakpoints.large}) {
    background: linear-gradient(
        to right,
        rgba(0, 0, 0, 1) 0%,
        rgba(0, 0, 0, 1) 20%,
        rgba(255, 255, 255, 0.3) 100%
      ),
      linear-gradient(
        to bottom,
        rgba(0, 0, 0, 1) 0%,
        rgba(255, 255, 255, 0.2) 100%
      );
  }
`;

const Nav = () => (
  <User>
    {({ data: { me: loggedInUser } }) => (
      <NavStyles data-test="nav">
        <TopBar>
          {loggedInUser && (
            <>
              <Link href="/sell">
                <a>Sell</a>
              </Link>
              <Link href="/orders">
                <a>Orders</a>
              </Link>
              <Link href="/me">
                <a>Account</a>
              </Link>
              <SignoutButton>Sign out</SignoutButton>
            </>
          )}
          {!loggedInUser && (
            <Link href="/signup">
              <a>Sign In</a>
            </Link>
          )}
        </TopBar>
        <BottomBar>
          <Link href="/items">
            <a>Shop</a>
          </Link>
          <Search />
          {loggedInUser && (
            <Mutation mutation={TOGGLE_CART_MUTATION}>
              {toggleCart => (
                <button onClick={toggleCart}>
                  My cart
                  <CartCount
                    count={loggedInUser.cart.reduce((tally, cartItem) => {
                      return tally + cartItem.quantity;
                    }, 0)}
                  />
                </button>
              )}
            </Mutation>
          )}
        </BottomBar>
      </NavStyles>
    )}
  </User>
);

export default Nav;
