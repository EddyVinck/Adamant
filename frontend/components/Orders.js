import React, { Component } from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import { format, formatDistance } from "date-fns";
import styled from "styled-components";
import formatMoney from "../lib/formatMoney";
import Link from "next/link";
import OrderItemStyles from "./styles/OrderItemStyles";
import Error from "./ErrorMessage";

const USER_ORDERS_QUERY = gql`
  query USER_ORDERS_QUERY {
    orders(orderBy: createdAt_DESC) {
      id
      createdAt
      total
      user {
        id
      }
      items {
        id
        title
        description
        image
        price
        quantity
      }
    }
  }
`;

const OrderOl = styled.ol`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

export default class Orders extends Component {
  render() {
    return (
      <Query query={USER_ORDERS_QUERY}>
        {({ data, loading, error }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>loading</p>;
          const orders = data.orders;
          console.log(data);
          return (
            <div>
              <h1>Orders</h1>
              <p>You have placed {orders.length} orders in the past.</p>
              <OrderOl>
                {orders.map(order => {
                  //
                  return (
                    <OrderItemStyles key={order.id} className="order">
                      <Link
                        href={{
                          pathname: "/order",
                          query: { id: order.id }
                        }}
                      >
                        <a>
                          <h2>
                            {format(order.createdAt, "MMMM d, YYYY h:mm a")}
                          </h2>
                          <div className="order-meta">
                            <p>
                              {order.items.reduce((a, b) => {
                                return a + b.quantity;
                              }, 0)}{" "}
                              items.
                            </p>
                            <p>{formatDistance(order.createdAt, new Date())}</p>
                            <p>Items:</p>
                            <p>{formatMoney(order.total)}</p>
                          </div>
                          <div className="images">
                            {order.items.map(item => (
                              <img
                                key={item.id}
                                src={item.image}
                                alt={item.title}
                              />
                            ))}
                          </div>
                        </a>
                      </Link>
                    </OrderItemStyles>
                  );
                })}
              </OrderOl>
            </div>
          );
        }}
      </Query>
    );
  }
}
