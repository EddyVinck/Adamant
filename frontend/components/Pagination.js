import React from "react";
import PaginationStyles from "./styles/PaginationStyles";
import Link from "next/link";
import Head from "next/head";

const Pagination = props => (
  <PaginationStyles data-test="pagination">
    <Head>
      <title>
        Adamant | Page {props.currentPage} of {props.pagesCount}
      </title>
    </Head>
    <Link prefetch href={props.prevPage}>
      <a className="prev" aria-disabled={props.prevDisabled}>
        ⬅ Prev
      </a>
    </Link>
    <p>
      Page {props.currentPage} of{" "}
      <span className="totalPages">{props.pagesCount}</span> pages.
    </p>
    <Link prefetch href={props.nextPage}>
      <a className="next" aria-disabled={props.nextDisabled}>
        Next ➡
      </a>
    </Link>
  </PaginationStyles>
);

export default Pagination;
