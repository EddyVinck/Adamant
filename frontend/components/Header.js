import styled from "styled-components";
import Link from "next/link";
import Nav from "./Nav";
import Router from "next/router";
import NProgress from "nprogress";
import Cart from "./Cart";

Router.onRouteChangeStart = () => {
  NProgress.start();
};
Router.onRouteChangeComplete = () => {
  NProgress.done();
};
Router.onRouteChangeError = () => {
  NProgress.done();
};

const StyledHeader = styled.header`
  .bar {
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: stretch;
    background: black;

    display: block;
    margin: 0 auto;

    @media (max-width: 1300px) {
      grid-template-columns: 1fr;
      justify-content: center;
    }
  }
`;

const Header = () => (
  <StyledHeader>
    <div className="bar">
      <Nav />
    </div>
    <Cart />
  </StyledHeader>
);

export default Header;
