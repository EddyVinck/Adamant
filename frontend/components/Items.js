import React, { Component } from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import styled from "styled-components";
import Item from "./Item";
import Pagination from "./Pagination";
import ItemPaginationData from "./ItemPaginationData";
import { perPage as itemsPerPage } from "../config";

const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${itemsPerPage}) {
    items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: 18px;
  margin: 0 auto;

  @media (min-width: ${props => props.theme.breakpoints.medium}) {
    grid-template-columns: 1fr 1fr;
  }
  @media (min-width: ${props => props.theme.breakpoints.large}) {
    grid-template-columns: 1fr 1fr 1fr;
    grid-gap: 24px;
    max-width: ${props => props.theme.maxWidth};
  }
`;

class Items extends Component {
  render() {
    const page = this.props.page;
    return (
      <div>
        <h1>All products</h1>
        <ItemPaginationData page={page}>
          {pagination => {
            return (
              <>
                <p>
                  {pagination.itemCount} products total | sorting by{" "}
                  <strong>most recent</strong>
                </p>
                <Pagination {...pagination} />
                <Query
                  query={ALL_ITEMS_QUERY}
                  // fetchPolicy="network-only" -> cache is unused
                  variables={{
                    skip: this.props.page * itemsPerPage - itemsPerPage,
                    first: itemsPerPage
                  }}
                >
                  {({ data, error, loading }) => {
                    if (loading) return <p>loading...</p>;
                    if (error) return <p>Error: {error.message}</p>;
                    return (
                      <ItemsList>
                        {data.items.map(item => (
                          <Item item={item} key={item.id} />
                        ))}
                      </ItemsList>
                    );
                  }}
                </Query>
                <Pagination {...pagination} />
              </>
            );
          }}
        </ItemPaginationData>
      </div>
    );
  }
}

export default Items;
export { ALL_ITEMS_QUERY };
