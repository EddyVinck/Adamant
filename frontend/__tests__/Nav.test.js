import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { ThemeProvider } from "styled-components";
import { theme } from "../components/Page";
import { CURRENT_USER_QUERY } from "../components/User";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeUser, fakeCartItem } from "../lib/testUtils";
import Nav from "../components/Nav";

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } }
  }
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } }
  }
];

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()]
        }
      }
    }
  }
];

describe("<Nav />", () => {
  it("renders a minimal nav when signed out", async () => {
    const wrapper = mount(
      <ThemeProvider theme={theme}>
        <MockedProvider mocks={notSignedInMocks}>
          <Nav />
        </MockedProvider>
      </ThemeProvider>
    );
    await wait();
    wrapper.update();
    // console.log(wrapper.debug());
    const nav = wrapper.find('nav[data-test="nav"]');
    expect(nav.exists()).toEqual(true);
    expect(toJSON(nav)).toMatchSnapshot();
  });

  it("renders full nav when signed in", async () => {
    const wrapper = mount(
      <ThemeProvider theme={theme}>
        <MockedProvider mocks={signedInMocks}>
          <Nav />
        </MockedProvider>
      </ThemeProvider>
    );
    await wait();
    wrapper.update();
    const nav = wrapper.find('nav[data-test="nav"]');
    expect(nav.exists()).toEqual(true);
    // console.log(nav.debug());
    expect(nav.children().length).toBe(2);
    expect(nav.text()).toMatch(/sign out/i);
    // console.log(nav.debug());
    // expect(toJSON(nav)).toMatchSnapshot();
  });

  it("renders the amount of items in the cart", async () => {
    const wrapper = mount(
      <ThemeProvider theme={theme}>
        <MockedProvider mocks={signedInMocksWithCartItems}>
          <Nav />
        </MockedProvider>
      </ThemeProvider>
    );
    await wait();
    wrapper.update();
    const nav = wrapper.find('nav[data-test="nav"]');
    expect(nav.exists()).toEqual(true);
    const count = nav.find("div.count");
    expect(count.text()).toBe("9"); // every fake item has a quantity of 3
    expect(toJSON(count)).toMatchSnapshot();
    // console.log(count.debug());
    // expect(toJSON(nav)).toMatchSnapshot();
  });
});
