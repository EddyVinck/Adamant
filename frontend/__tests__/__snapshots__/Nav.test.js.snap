// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders a minimal nav when signed out 1`] = `
<nav
  className="NavStyles-sc-11c0d2g-0 hoyqJE"
  data-test="nav"
>
  <Nav__TopBar>
    <Nav__BottomBar
      className="Nav__TopBar-sc-1h4lqeb-1 eBFvTI"
    >
      <div
        className="Nav__TopBar-sc-1h4lqeb-1 eBFvTI Nav__BottomBar-sc-1h4lqeb-0 dFJmLJ"
      >
        <Link
          href="/signup"
        >
          <a
            href="/signup"
            onClick={[Function]}
          >
            Sign In
          </a>
        </Link>
      </div>
    </Nav__BottomBar>
  </Nav__TopBar>
  <Nav__BottomBar>
    <div
      className="Nav__BottomBar-sc-1h4lqeb-0 dFJmLJ"
    >
      <Link
        href="/items"
      >
        <a
          href="/items"
          onClick={[Function]}
        >
          Shop
        </a>
      </Link>
      <AutoComplete>
        <DropDown__SearchStyles>
          <div
            className="DropDown__SearchStyles-n5d97g-2 yMPcx"
          >
            <Downshift
              defaultHighlightedIndex={null}
              defaultInputValue=""
              defaultIsOpen={false}
              defaultSelectedItem={null}
              environment={[Window]}
              getA11yStatusMessage={[Function]}
              itemToString={[Function]}
              onChange={[Function]}
              onInputValueChange={[Function]}
              onOuterClick={[Function]}
              onSelect={[Function]}
              onStateChange={[Function]}
              onUserAction={[Function]}
              scrollIntoView={[Function]}
              selectedItemChanged={[Function]}
              stateReducer={[Function]}
              suppressRefError={false}
            >
              <div
                aria-expanded={false}
                aria-haspopup="listbox"
                aria-labelledby="downshift-0-label"
                aria-owns={null}
                role="combobox"
              >
                <ApolloConsumer>
                  <input
                    aria-activedescendant={null}
                    aria-autocomplete="list"
                    aria-controls={null}
                    aria-labelledby="downshift-0-label"
                    autoComplete="off"
                    className=""
                    id="search"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Search for an item.."
                    type="search"
                    value=""
                  />
                </ApolloConsumer>
              </div>
            </Downshift>
          </div>
        </DropDown__SearchStyles>
      </AutoComplete>
    </div>
  </Nav__BottomBar>
</nav>
`;

exports[`<Nav /> renders the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-1 kvCVqE"
>
  9
</div>
`;
